homeassistant/components/switchbot_cloud/fan.py

```diff
@@ -0,0 +1,138 @@
+
+from . import SwitchbotCloudData
+from .const import DOMAIN
+from .entity import SwitchBotCloudEntity
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config: ConfigEntry,
+    async_add_entities: AddConfigEntryEntitiesCallback,
+) -> None:
+    """Set up SwitchBot Cloud entry."""
+    data: SwitchbotCloudData = hass.data[DOMAIN][config.entry_id]
+    async_add_entities(
+        SwitchBotCloudFan(data.api, device, coordinator)
+        for device, coordinator in data.devices.fans
+    )
+
+
+class SwitchBotCloudFan(SwitchBotCloudEntity, FanEntity):
+    """Representation of a SwitchBot Battery Circulator Fan."""
+
+    _attr_name = None
+
+    _attr_supported_features = (
+        FanEntityFeature.SET_SPEED
+        | FanEntityFeature.PRESET_MODE
+        | FanEntityFeature.TURN_OFF
+        | FanEntityFeature.TURN_ON
+    )
+    _attr_preset_modes = list(BatteryCirculatorFanMode)
+
+    _attr_is_on: bool | None = None
+
+    @property
+    def is_on(self) -> bool | None:
+        """Return true if the entity is on."""
+        return self._attr_is_on
+
+    def _set_attributes(self) -> None:
+        """Set attributes from coordinator data."""
+        if self.coordinator.data is None:
+            return
+
+        power: str = self.coordinator.data["power"]
+        mode: str = self.coordinator.data["mode"]
+        fan_speed: str = self.coordinator.data["fanSpeed"]
+        self._attr_is_on = power == "on"
+        self.preset_mode = mode
+        self.percentage = int(fan_speed)
+        if self.is_on and mode == BatteryCirculatorFanMode.DIRECT.value:
+            self._attr_supported_features = (
+                FanEntityFeature.SET_SPEED
+                | FanEntityFeature.PRESET_MODE
+                | FanEntityFeature.TURN_OFF
+                | FanEntityFeature.TURN_ON
+            )
+        else:
+            self._attr_supported_features = (
+                FanEntityFeature.PRESET_MODE
+                | FanEntityFeature.TURN_OFF
+                | FanEntityFeature.TURN_ON
+            )
+
+    async def async_turn_on(
+        self,
+        percentage: int | None = None,
+        preset_mode: str | None = None,
+        **kwargs: Any,
+    ) -> None:
+        """Turn on the fan."""
+        await self.send_api_command(CommonCommands.ON)
+        await self.send_api_command(
+            command=BatteryCirculatorFanCommands.SET_WIND_MODE,
+            parameters=str(self.preset_mode),
+        )
+        if self.preset_mode == BatteryCirculatorFanMode.DIRECT.value:
+            await self.send_api_command(
+                command=BatteryCirculatorFanCommands.SET_WIND_SPEED,
+                parameters=str(self.percentage),
+            )
+        await self.coordinator.async_refresh()
+        await asyncio.sleep(2)
+        self.async_write_ha_state()
+        await asyncio.sleep(2)
+        await self.coordinator.async_refresh()
```

Comment:
Hold up, why do we need both now? This now means that we send the update, then update the state (which writes the state), then we wait 2 seconds, write state (which hasn't changed), then wait another 2 seconds, and then request data again?

With my previous comment I meant more like, it's usually either one or the other. And if I am missing something, please let me know :)

Reply:
my  mistake . I set like below , I tested many time , set delay =5s have the best effects 
```
await asyncio.sleep(5)
await self.coordinator.async_refresh()
```